teamcity: failed test: gossip/restart

The following tests appear to have failed on master (roachtest): acceptance/gossip/restart

You may want to check [for open issues](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+is%3Aopen+gossip/restart).

[#1459377](https://teamcity.cockroachdb.com/viewLog.html?buildId=1459377):


```
acceptance/gossip/restart
--- FAIL: roachtest/acceptance/gossip/restart (43.098s)
test artifacts and logs in: artifacts/acceptance/gossip/restart/run_1
	gossip.go:201,gossip.go:286,acceptance.go:69,test_runner.go:673: failed to get gossip status from node 4: Get http://127.0.0.1:26264/_status/gossip/local: dial tcp 127.0.0.1:26264: connect: connection refused

test artifacts and logs in: artifacts/acceptance/gossip/restart/run_1
	gossip.go:201,gossip.go:286,acceptance.go:69,test_runner.go:673: failed to get gossip status from node 4: Get http://127.0.0.1:26264/_status/gossip/local: dial tcp 127.0.0.1:26264: connect: connection refused





```

Please assign, take a look and update the issue accordingly.

--- a/pkg/storage/batcheval/cmd_lease.go
+++ b/pkg/storage/batcheval/cmd_lease.go
@@ -58,7 +58,15 @@ func checkCanReceiveLease(rec EvalContext) error {
 		//
 		// Since the leaseholder can't remove itself and is a VOTER_FULL, we
 		// also know that in any configuration there's at least one VOTER_FULL.
-		return errors.Errorf(`cannot transfer lease to replica of type %s`, t)
+		//
+		// TODO(tbg): if this code path is hit during a lease transfer (we check
+		// upstream of raft, but this check has false negatives) then we are in
+		// a situation where the leaseholder is a node that has set its
+		// minProposedTS and won't be using its lease any more. Either the setting
+		// of minProposedTS needs to be "reversible" (tricky) or we make the
+		// lease evaluation succeed, though with a lease that's "invalid" so that
+		// a new lease can be requested right after.
+		return errors.Errorf(`replica of type %s cannot hold lease`, t)
 	}
 	return nil
 }

--- a/pkg/storage/batcheval/cmd_lease_request.go
+++ b/pkg/storage/batcheval/cmd_lease_request.go
@@ -36,6 +36,12 @@ func RequestLease(
 	// newFailedLeaseTrigger() to satisfy stats.
 	args := cArgs.Args.(*roachpb.RequestLeaseRequest)
 
+	prevLease, _ := cArgs.EvalCtx.GetLease()
+	rErr := &roachpb.LeaseRejectedError{
+		Existing:  prevLease,
+		Requested: args.Lease,
+	}
+
 	// For now, don't allow replicas of type LEARNER to be leaseholders. There's
 	// no reason this wouldn't work in principle, but it seems inadvisable. In
 	// particular, learners can't become raft leaders, so we wouldn't be able to
@@ -48,14 +54,8 @@ func RequestLease(
 	// If this check is removed at some point, the filtering of learners on the
 	// sending side would have to be removed as well.
 	if err := checkCanReceiveLease(cArgs.EvalCtx); err != nil {
-		return newFailedLeaseTrigger(false /* isTransfer */), err
-	}
-
-	prevLease, _ := cArgs.EvalCtx.GetLease()
-
-	rErr := &roachpb.LeaseRejectedError{
-		Existing:  prevLease,
-		Requested: args.Lease,
+		rErr.Message = err.Error()
+		return newFailedLeaseTrigger(false /* isTransfer */), rErr
 	}
 
 	// MIGRATION(tschottdorf): needed to apply Raft commands which got proposed

--- a/pkg/storage/batcheval/cmd_lease_test.go
+++ b/pkg/storage/batcheval/cmd_lease_test.go
@@ -129,9 +129,13 @@ func TestLeaseCommandLearnerReplica(t *testing.T) {
 	// Learners are not allowed to become leaseholders for now, see the comments
 	// in TransferLease and RequestLease.
 	_, err := TransferLease(ctx, nil, cArgs, nil)
-	require.EqualError(t, err, `cannot transfer lease to replica of type LEARNER`)
+	require.EqualError(t, err, `replica of type LEARNER cannot hold lease`)
 
 	cArgs.Args = &roachpb.RequestLeaseRequest{}
 	_, err = RequestLease(ctx, nil, cArgs, nil)
-	require.EqualError(t, err, `cannot transfer lease to replica of type LEARNER`)
+
+	const exp = `cannot replace lease repl=(n0,s0):? seq=0 start=0.000000000,0 exp=<nil> ` +
+		`with repl=(n0,s0):? seq=0 start=0.000000000,0 exp=<nil>: ` +
+		`replica of type LEARNER cannot hold lease`
+	require.EqualError(t, err, exp)
 }

--- a/pkg/storage/batcheval/cmd_resolve_intent_test.go
+++ b/pkg/storage/batcheval/cmd_resolve_intent_test.go
@@ -42,6 +42,7 @@ type mockEvalCtx struct {
 	gcThreshold      hlc.Timestamp
 	term, firstIndex uint64
 	canCreateTxnFn   func() (bool, hlc.Timestamp, roachpb.TransactionAbortedReason)
+	lease            roachpb.Lease
 }
 
 func (m *mockEvalCtx) String() string {
@@ -119,7 +120,7 @@ func (m *mockEvalCtx) GetLastReplicaGCTimestamp(context.Context) (hlc.Timestamp,
 	panic("unimplemented")
 }
 func (m *mockEvalCtx) GetLease() (roachpb.Lease, roachpb.Lease) {
-	panic("unimplemented")
+	return m.lease, roachpb.Lease{}
 }
 
 func TestDeclareKeysResolveIntent(t *testing.T) {

